### PR TITLE
SOLR-15951: Local packages

### PR DESCRIPTION
**Pull request by chatman**
https://issues.apache.org/jira/browse/SOLR-15951

Ability to load packages from the filesystem, useful for first party packages.

TODO: Generate the local_packages.json on the fly (based on the Solr's version that is being built).

### CODE REVIEW

File: solr/contrib/scripting/src/java/org/apache/solr/scripting/update/ScriptUpdateProcessorFactory.java [hunk 1]
```diff
@@ -49,11 +49,7 @@
 import java.security.PrivilegedActionException;
 import java.security.PrivilegedExceptionAction;
 import java.security.ProtectionDomain;
-import java.util.Set;
-import java.util.LinkedHashSet;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Collection;
+import java.util.*;
```
Comment (HoustonPutman) [inline solr/contrib/scripting/src/java/org/apache/solr/scripting/update/ScriptUpdateProcessorFactory.java:52]: please leave these separate.

File: solr/core/src/java/org/apache/solr/pkg/PackageLoader.java [hunk 2]
```diff
@@ -52,10 +56,14 @@
 public class PackageLoader implements Closeable {
   private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
   public static final String LATEST = "$LATEST";
-
+  public static final String LOCAL_PACKAGES_DIR = System.getProperty("solr.packages.local.dir");
+  public static final String LOCAL_PACKAGES_WHITELIST = System.getProperty("solr.packages.local.whitelist", "");
```
Comment (HoustonPutman) [inline solr/core/src/java/org/apache/solr/pkg/PackageLoader.java:60]: instead of `whitelist`, how about `names`, 'use' or `enable`. We are steering away from `whitelist`/`blacklist` now, and it doesn't work as well as other options.

File: solr/core/src/test/org/apache/solr/pkg/TestPackages.java [hunk 3]
```diff
@@ -61,6 +61,7 @@
 import org.junit.Before;
 import org.junit.Test;
 
+import java.io.File;
```
Comment (HoustonPutman) [inline solr/core/src/test/org/apache/solr/pkg/TestPackages.java:64]: is this needed?

File: solr/packaging/build.gradle [hunk 4]
```diff
@@ -86,12 +86,20 @@ distributions {
         include "NOTICE.txt"
       })
 
+      from(project(":solr:contrib").projectDir, {
```
Comment (HoustonPutman) [inline solr/packaging/build.gradle:89]: the `modules` directory should have a build.gradle that creates the `local_packages.json` in it's build directory, then puts that into a gradle configuration. Then this file can refer to that configuration.

same with `local_packages.json` and `manifest.json` below.

File: gradle/solr/packaging.gradle [hunk 5]
```diff
@@ -60,7 +60,11 @@ configure(allprojects.findAll {project -> project.path.startsWith(":solr:contrib
     // An aggregate that configures lib and test-lib in a temporary location.
     task assemblePackaging(type: Sync) {
       from "README.md"
-      
+
+      from "manifest.json"
+
+      from "src/resources/solr-default-tika-config.xml"
```
Comment (HoustonPutman) [inline gradle/solr/packaging.gradle:66]: You can put this in the `extraction` module's build.gradle using

```
assemblePackaging {
  from "src/resources/solr-default-tika-config.xml"
}
```

File: solr/bin/solr [hunk 6]
```diff
@@ -2264,6 +2264,7 @@ function start_solr() {
     # users who don't care about useful error msgs can override in SOLR_OPTS with +OmitStackTraceInFastThrow
     "${SOLR_HOST_ARG[@]}" "-Duser.timezone=$SOLR_TIMEZONE" "-XX:-OmitStackTraceInFastThrow" \
     "-XX:OnOutOfMemoryError=$SOLR_TIP/bin/oom_solr.sh $SOLR_PORT $SOLR_LOGS_DIR" \
+    "-Dsolr.packages.local.dir=$SOLR_TIP/contrib" \
```
Comment (HoustonPutman) [inline solr/bin/solr:2267]: This should be in the windows `.cmd` as well.

File: solr/packaging/build.gradle [hunk 7]
```diff
@@ -86,12 +86,20 @@ distributions {
...
+      })
+
+
       from(project(":solr").projectDir, {
         include "bin/**"
         include "licenses/**"
         exclude "licenses/README.committers.txt"
         include "CHANGES.txt"
         include "README.md"
+        include "local_packages.json"
```
Comment (HoustonPutman) [inline solr/packaging/build.gradle:101]: These two files aren't under `/solr`, they are in the individual module directories...

File: solr/core/src/java/org/apache/solr/pkg/PackageLoader.java [hunk 8]
```diff
@@ -70,17 +78,49 @@
...
+    }
   }
 
   public PackageAPI getPackageAPI() {
     return packageAPI;
   }
 
   public Package getPackage(String key) {
-    return packageClassLoaders.get(key);
+   return packageClassLoaders.get(key);
```
Comment (HoustonPutman) [inline solr/core/src/java/org/apache/solr/pkg/PackageLoader.java:123]: wrong indentation.

**Review by HoustonPutman — CHANGES_REQUESTED** [review]
This is by no means a thorough review. Just initial things that stand out to me.

Also as you said this needs to auto-generate these `manifest.json` and `local_packages.json` files. This can't be committed until that is happening.

Also how are these 1st party packages going to determine version-constraints? Should they even have them, given that users have installed them already or they are provided with the solr distribution?

File: solr/core/src/test/org/apache/solr/pkg/TestLocalPackages.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    pkgVersion.pkg = PKG_NAME;
+    p.packages.put(PKG_NAME, Collections.singletonList(pkgVersion));
+
+    log.info("local_packages.json: " + Utils.toJSONString(p));
+    log.info("Local packages dir: " + localPackagesDir);
+
+    MiniSolrCloudCluster cluster =
+            configureCluster(4)
+                    .withJettyConfig(builder -> builder.enableV2(true))
+                    .withJettyConfig(it -> it.withPreStartupHook(jsr -> {
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/pkg/TestLocalPackages.java:55]: I'm curious; why is this needed vs simply calling this code before cluster is instantiated?
Also, I think it's better style to put all this code into a method and simply call it rather than have a long lambda.

File: solr/core/src/test/org/apache/solr/pkg/TestLocalPackages.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+                      try {
+                        File pkgDir = new File(localPackagesDir);
+                        if(!pkgDir.exists()) {
+                          pkgDir.mkdir();
+                        }
+                        try (FileInputStream fis = new FileInputStream(getFile("runtimecode/runtimelibs.jar.bin"))) {
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/pkg/TestLocalPackages.java:61]: I could be wrong but I bet there are simpler / more concise idioms available to us in Java 11 to do this basic task.

File: solr/core/src/java/org/apache/solr/pkg/PackageLoader.java [hunk 11]
```diff
@@ -115,6 +120,36 @@ private void loadLocalPackages(List<String> enabledPackages) {
...
 
+  /**If a directory with no local_packages.json is provided,
+   * each sub directory that contains one or more jar files
+   * will be treated as a package and each jar file in that
+   * directory is added to the package classpath
+   */
+  private PackageAPI.Packages readDirAsPackages(Path packagesPath) {
+    PackageAPI.Packages localDirAsPackages = new PackageAPI.Packages();
+    packagesPath.toFile().list((dir, pkgName) -> {
+      File subDir = new File(dir, pkgName);
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/pkg/PackageLoader.java:131]: Let's not use java.io.File and instead use Path, wherever possible.

File: solr/core/src/test/org/apache/solr/pkg/TestLocalPackages.java [hunk 12]
```diff
@@ -12,33 +12,75 @@
...
   private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  public void testLocalPackagesAsDir() throws Exception {
+    String PKG_NAME = "mypkg";
+    String jarName = "mypkg1.jar";
+    String COLLECTION_NAME = "testLocalPkgsColl";
+    String localPackagesDir = "testpkgdir";
+    System.setProperty(SOLR_PACKAGES_LOCAL_ENABLED, PKG_NAME);
+    System.setProperty(LPD, localPackagesDir);
+    MiniSolrCloudCluster cluster =
+            configureCluster(4)
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/pkg/TestLocalPackages.java:31]: Do we need a cluster this big to test this?

File: solr/core/src/java/org/apache/solr/pkg/PackageLoader.java [hunk 13]
```diff
@@ -56,9 +52,12 @@
 public class PackageLoader implements Closeable {
   private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
   public static final String LATEST = "$LATEST";
-  public static final String LOCAL_PACKAGES_DIR = System.getProperty("solr.packages.local.dir");
-  public static final String LOCAL_PACKAGES_WHITELIST = System.getProperty("solr.packages.local.whitelist", "");
+  public static final String LPD = "solr.packages.local.dir";
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/pkg/PackageLoader.java:55]: Expand the acronym please.  I didn't know what it was at the spot that referenced it.

**Review by dsmiley — COMMENTED** [review]
I look forward to trying this!

File: gradle/solr/packaging.gradle (hunk 5 above)
Comment (HoustonPutman) [reply gradle/solr/packaging.gradle]: I removed this. The `solr-default-tika-config.xml` is already included in the JAR, given that it's in `src/resources`

File: solr/contrib/scripting/manifest.json [hunk 14]
```diff
@@ -0,0 +1,22 @@
+{
+    "version-constraint": "10",
```
Comment (HoustonPutman) [inline solr/contrib/scripting/manifest.json:2]: What is the point of version constraints for local packages?

Can this be removed altogether? If not, then it should be autogenerated by gradle when copying over to the `build/packaging` directory. It should not be hardcoded.

Comment (janhoy) [reply solr/contrib/scripting/manifest.json:2]: Agreed. Local package bug fixes are always released together with Solr, so you'd never want to, say, run Solr 9.0.0 with a LTR package 9.1.0. Makes no sense. So I think we could just remove version-constraint for these manifests, if that it allowed.

File: solr/contrib/build.gradle [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+task createLocalPackagesJson {
+  dependsOn configurations.modulePackages
+
+  doLast {
+    def packagesJson = [
+        packages: [:]
+    ]
+    project(":solr:contrib:scripting").configurations.packaging.resolve().each {
+      logger.error("It: $it")
+    }
```
Comment (madrob) [inline solr/contrib/build.gradle:54]: vestigal logging?

File: solr/contrib/build.gradle [hunk 16]
```diff
@@ -0,0 +1,77 @@
+import groovy.json.JsonOutput
+import org.gradle.api.internal.artifacts.dependencies.DefaultProjectDependency
```
Comment (madrob) [inline solr/contrib/build.gradle:2]: unused?

Comment (HoustonPutman) [reply solr/contrib/build.gradle:2]: yeah forgot to remove all the debugging stuff.

File: solr/core/src/java/org/apache/solr/pkg/PackageLoader.java [hunk 17]
```diff
@@ -274,20 +349,25 @@ public void writeMap(EntryWriter ew) throws IOException {
...
-          throw new RuntimeException("Cannot load package: " +errs);
-        }
-        for (String file : version.files) {
-          paths.add(coreContainer.getPackageStoreAPI().getPackageStore().getRealpath(file));
+        if(localPkgDir != null) {
+          for (String file : v.files) {
+            if(file.charAt(0)== '/') file =file.substring(1);
+            paths.add( localPkgDir.resolve(file).toAbsolutePath()) ;
+          }
+        } else {
```
Comment (janhoy) [inline solr/core/src/java/org/apache/solr/pkg/PackageLoader.java:361]: Some code smell here with if-then-else based on quite different behaviour - local packages that are just added t path vs remote packages that need validation etc. Can we have sublcassing instead, either `ValidatingPackageLoader` and `LocalPackageLoader`, and the same for Package.Version?

File: solr/core/src/test/org/apache/solr/pkg/TestLocalPackages.java [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+    pkgVersion.files = Collections.singletonList(jarName);
+    pkgVersion.version = "0.1";
+    pkgVersion.pkg = PKG_NAME;
+    p.packages.put(PKG_NAME, Collections.singletonList(pkgVersion));
+
+    log.info("local_packages.json: " + Utils.toJSONString(p));
+    log.info("Local packages dir: " + localPackagesDir);
+    System.setProperty(SOLR_PACKAGES_LOCAL_ENABLED, PKG_NAME);
+    System.setProperty(LPD, localPackagesDir);
+    MiniSolrCloudCluster cluster =
```
Comment (janhoy) [inline solr/core/src/test/org/apache/solr/pkg/TestLocalPackages.java:95]: Wrap cluster creation in a try-finally block?

File: solr/core/src/java/org/apache/solr/pkg/PackageLoader.java (hunk 17 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/pkg/PackageLoader.java:361]: Yeah, I think a subclass with a different name would be better

File: solr/core/src/java/org/apache/solr/pkg/PackageLoader.java [hunk 19]
```diff
@@ -70,17 +77,85 @@
...
+      packageClassLoaders.put(e.getKey(), p);
+    }
+  }
+
+  /**If a directory with no local_packages.json is provided,
+   * each sub directory that contains one or more jar files
+   * will be treated as a package and each jar file in that
+   * directory is added to the package classpath
+   */
+  private PackageAPI.Packages readDirAsPackages(Path packagesPath) {
```
[the remark this answers is not in the document]
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/pkg/PackageLoader.java:128]: `TestLocalPackages#testLocalPackagesAsDir()`

File: solr/core/src/test/org/apache/solr/pkg/TestLocalPackages.java (hunk 9 above)
Comment (noblepaul) [reply solr/core/src/test/org/apache/solr/pkg/TestLocalPackages.java:98]: > I'm curious; why is this needed vs simply calling this code before cluster is instantiated?

We need this code to be executed after the `JettySolrRunner` is created but before the jetty is started because the paths are only available at that point.

> I think it's better style to put all this code into a method and simply call it rather than have a long lambda.

👍

File: solr/core/src/test/org/apache/solr/pkg/TestLocalPackages.java (hunk 12 above)
Comment (noblepaul) [reply solr/core/src/test/org/apache/solr/pkg/TestLocalPackages.java:32]: 2 should be fine

File: solr/core/src/test/org/apache/solr/pkg/TestLocalPackages.java (hunk 10 above)
Comment (noblepaul) [reply solr/core/src/test/org/apache/solr/pkg/TestLocalPackages.java:105]: sure. As it was a test, I didn't explore the idiomatic way of doing this

File: solr/core/src/java/org/apache/solr/pkg/PackageLoader.java [hunk 20]
```diff
@@ -78,29 +82,31 @@
...
-            localPkgsDir :
-            coreContainer.getSolrHome()+ File.separator+ localPkgsDir).toPath();
+  private void loadLocalPackages(String localPkgsDir,  List<String> enabledPackages) {
+    final Path packagesPath = localPkgsDir.charAt(0) == File.separatorChar ?
+            Paths.get(localPkgsDir) :
+            Paths.get(coreContainer.getSolrHome()).resolve(localPkgsDir);
     log.info("Packages to be loaded from directory: {}", packagesPath);
 
-    if (!packagesPath.toFile().exists()) {
+    if (!Files.exists(packagesPath)) {
```
Comment (janhoy) [inline solr/core/src/java/org/apache/solr/pkg/PackageLoader.java:102]: Is there or will there be another way than sysProp to set the local packages dir? If it can be set by some api, it should be checked against allowPaths in cc.

File: solr/bin/solr (hunk 6 above)
Comment (janhoy) [reply solr/bin/solr:2267]: Should not the bundled "modules" folder always be available? Is it even necessary to make the local packages dir customizable like this? I mean, if you perhaps have your own inhouse pkg that you would want to load in a "trusted" way, you would either copy it into the "modules" folder or you'd want pkg mgr to resolve some other location **in addition**, not instead of the 1st party pkgs?

PS: Is it time to merge in latest master to cut everything over to "modules"?

File: solr/core/src/java/org/apache/solr/pkg/PackageLoader.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (janhoy) [reply solr/core/src/java/org/apache/solr/pkg/PackageLoader.java:135]: Well, that is a happy-path integration test. I more thought of a unit test that simply creates a tmpDirectory folder with several different (and erraneous) sub folders inside to validate that you get exactley the expected packages back, and that failure cases are handled correctly.

File: solr/core/src/java/org/apache/solr/pkg/PackageLoader.java (hunk 20 above)
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/pkg/PackageLoader.java:102]: Sys prop is the only way to do it now. We should NOT have an API to do this because this also goes against the `immutable deployments` narrative . However we may, in the future,  use other startup mechanisms to configure the same (say `solr.xml`)

File: solr/core/src/java/org/apache/solr/pkg/PackageLoader.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (noblepaul) [reply solr/core/src/java/org/apache/solr/pkg/PackageLoader.java:135]: I shall add one. but the code is written to handle directories with jars only

**Comment by chatman** [comment]
I had some issues rebasing to latest main around the gradle part (due to the "contrib" vs "modules" switch). I'll give it a try again shortly. FYI @noblepaul.

**Comment by chatman** [comment]
@HoustonPutman Gradle changes look and work great :-)

**Comment by chatman** [comment]
@HoustonPutman I tried to merge main into this branch, rebase on main, as well as apply these changes on a fresh branch forked from main. 

On every occasion, the doLast task in solr/modules/build.gradle is not executing at all, and hence the local_packages.json isn't getting generated. Here's a branch based on fresh main. https://github.com/apache/solr/tree/jira/solr-15951. Any idea, please?

**Comment by HoustonPutman** [comment]
Ok I merged `main` in and got stuff working. Haven't really tested everything, but I did test that the packaging works, and the the `local_packages.json` and the `manifest.json` files get created.

File: solr/bin/solr (hunk 6 above)
Comment (HoustonPutman) [reply solr/bin/solr]: So I agree that this likely doesn't need to be customizable. But if we intend to allow customization, how about we just provide this as a default when reading in the value? (And also read from both sysProp as well as envVar). That way there's no need to modify the bin/solr or bin/solr.cmd
